pkg/domain/infra/abi/containers.go

```diff
@@ -1806,13 +1806,42 @@ func (ic *ContainerEngine) ContainerUpdate(ctx context.Context, updateOptions *e
 	if len(containers) != 1 {
 		return "", fmt.Errorf("container not found")
 	}
+	container := containers[0].Container
 
 	var restartPolicy *string
 	if updateOptions.Specgen.RestartPolicy != "" {
 		restartPolicy = &updateOptions.Specgen.RestartPolicy
 	}
 
-	if err = containers[0].Update(updateOptions.Specgen.ResourceLimits, restartPolicy, updateOptions.Specgen.RestartRetries); err != nil {
+	healthCheckConfig, changedHealthCheck, err := specgenutil.GetNewHealthCheckConfig(container.HealthCheckConfig(), *updateOptions.ChangedHealthCheckConfiguration)
+	if err != nil {
+		return "", err
+	}
+	if changedHealthCheck {
+		if err := container.UpdateHealthCheckConfig(healthCheckConfig); err != nil {
+			return "", err
+		}
+	}
+
+	startupHealthCheckConfig, changedStartupHealthCheck, err := specgenutil.GetNewStartupHealthCheckConfig(container.Config().StartupHealthCheckConfig, *updateOptions.ChangedHealthCheckConfiguration)
+	if err != nil {
+		return "", err
+	}
+	if changedStartupHealthCheck {
+		if err := container.UpdateStartupHealthCheckConfig(startupHealthCheckConfig); err != nil {
+			return "", err
+		}
+	}
+
+	globalHealthCheckOptions, err := specgenutil.GetNewGlobalHealthCheck(*updateOptions.ChangedHealthCheckConfiguration)
+	if err != nil {
+		return "", err
+	}
+	if err := container.UpdateGlobalHealthCheckConfig(globalHealthCheckOptions); err != nil {
+		return "", err
+	}
+
+	if err = container.Update(updateOptions.Specgen.ResourceLimits, restartPolicy, updateOptions.Specgen.RestartRetries); err != nil {
```

Comment:
this entire flow makes updates not atomic which seems bad. Every time you call into libpod we have to lock again. sync the state again and then do whatever we do save the db config and state again. Doing this multiple times is slow and not atomic which means ugly race conditions most likely.

I rather have this moved into one libpod function that acts under one lock.